pkg/graphviz/traceflow.go

```diff
@@ -0,0 +1,265 @@
+		}
+		labelStr += "\n" + string(o.Action)
+		if o.Component == opsv1alpha1.NetworkPolicy && len(o.NetworkPolicy) > 0 {
+			labelStr += "\nNetpol: " + o.NetworkPolicy
+		}
+		if o.Action == opsv1alpha1.Dropped {
+			node.SetColor("violet")
+		} else {
+			node.SetColor("turquoise3")
+			if len(o.TunnelDstIP) > 0 {
+				labelStr += "\nTo: " + o.TunnelDstIP
+			}
+		}
+		node.SetLabel(labelStr)
+	}
+	return nodes
+}
+
+func isSender(result opsv1alpha1.NodeResult) bool {
+	if len(result.Observations) == 0 {
+		return false
+	}
+	if result.Observations[0].Component != opsv1alpha1.SpoofGuard || result.Observations[0].Action != opsv1alpha1.Forwarded {
+		return false
+	}
+	return true
+}
+
+func isReceiver(result opsv1alpha1.NodeResult) bool {
+	if len(result.Observations) == 0 {
+		return false
+	}
+	if result.Observations[0].Component != opsv1alpha1.Forwarding || result.Observations[0].Action != opsv1alpha1.Received {
+		return false
+	}
+	return true
+}
+
+func getNodeResult(tf *opsv1alpha1.Traceflow, fn func(result opsv1alpha1.NodeResult) bool) *opsv1alpha1.NodeResult {
+	for _, result := range tf.Status.Results {
+		if fn(result) {
+			return &result
+		}
+	}
+	return nil
+}
+
+func getSrcNodeName(tf *opsv1alpha1.Traceflow) string {
+	if len(tf.Spec.Source.Namespace) > 0 && len(tf.Spec.Source.Pod) > 0 {
+		return tf.Spec.Source.Namespace + "/" + tf.Spec.Source.Pod
+	}
+	return ""
+}
+
+func getDstNodeName(tf *opsv1alpha1.Traceflow) string {
+	if len(tf.Spec.Destination.Namespace) > 0 && len(tf.Spec.Destination.Pod) > 0 {
+		return tf.Spec.Destination.Namespace + "/" + tf.Spec.Destination.Pod
+	}
+	return ""
+}
+
+// In Graphviz, clusters are surrounded by a pair of "{}" with string "subgraph ClusterName" before them.
+// The function finds the start and end index of specific cluster.
+func findClusterString(graphStr string, clusterName string) (startIndex int, endIndex int) {
+	startIndex = strings.Index(graphStr, "subgraph "+clusterName)
+	if startIndex == -1 {
+		return 0, 1
+	}
+	endIndex = startIndex
+	for graphStr[endIndex] != '{' {
+		endIndex++
+	}
+	// Depth represents the number of "{" minus the number of "}" from the start index of current cluster to endIndex.
+	// When depth is zero for the first time, it indicates that we successfully find the end index of the cluster.
+	depth := 1
+	for depth > 0 {
+		endIndex++
+		if graphStr[endIndex] == '{' {
+			depth++
+		}
+		if graphStr[endIndex] == '}' {
+			depth--
+		}
+	}
+	return startIndex, endIndex + 1
+}
+
+func genOutput(g *graphviz.Graphviz, graph *cgraph.Graph, isSingleCluster bool) string {
+	var buf bytes.Buffer
+	if err := g.Render(graph, "dot", &buf); err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("Graphz: \n%v", buf.String())
+	if err := graph.Close(); err != nil {
+		log.Fatal(err)
+	}
+	g.Close()
+
+	str := buf.String()
+	if isSingleCluster {
+		return str
+	}
+	// Swap source and destination cluster if destination cluster appears before source cluster.
```

Comment:
Maybe you can add some comments to explain what is a cluster?

Reply:
Comments added.